microservices/s3_to_redshift/s3_to_redshift_gov_assets_derived.py

```diff
@@ -0,0 +1,100 @@
+###################################################################
+# Script Name   : s3_to_redshift_gov_assets_derived.py
+#
+# Description   : Creates asset_downloads_derived, which is a
+#               : persistent derived table (PDT) 
+#
+# Requirements  : You must set the following environment variable
+#               : to establish credentials for the pgpass user microservice
+#
+#               : export pguser=<<database_username>>
+#               : export pgpass=<<database_password>>
+#
+#
+# Usage         : python s3_to_redshift_gov_assets_derived.py
+#
+
+import os
+import psycopg2
+import logging
+
+# Logging has two handlers: INFO to stdout and DEBUG to a file handler
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+handler = logging.StreamHandler()
+handler.setLevel(logging.INFO)
+formatter = logging.Formatter("%(message)s")
+handler.setFormatter(formatter)
+logger.addHandler(handler)
+
+if not os.path.exists('logs'):
+    os.makedirs('logs')
+
+log_filename = '{0}'.format(os.path.basename(__file__).replace('.py', '.log'))
+handler = logging.FileHandler(os.path.join('logs', log_filename),
+                              "a", encoding=None, delay="true")
+handler.setLevel(logging.DEBUG)
+formatter = logging.Formatter("%(levelname)s:%(name)s:%(asctime)s:%(message)s")
+handler.setFormatter(formatter)
+logger.addHandler(handler)
+
+conn_string = """
+dbname='{dbname}' host='{host}' port='{port}' user='{user}' password={password}
+""".format(dbname='snowplow',
+           host='redshift.analytics.gov.bc.ca',
+           port='5439',
+           user=os.environ['pguser'],
+           password=os.environ['pgpass'])
+
+query = '''
+    BEGIN;
+    SET SEARCH PATH TO cmslite;
+    DROP TABLE IF EXISTS asset_downloads_derived;
+    CREATE TABLE asset_downloads_derived AS
+    SELECT 'https://www2.gov.bc.ca' || replace(replace(assets.request_string,'GET ',''), ' HTTP/1.0','') AS asset_path,
+    assets.date_timestamp::TIMESTAMP,
```

Comment:
This is a `TIMESTAMP` so the offset is truncated when inserted into Redshift. See: https://docs.aws.amazon.com/redshift/latest/dg/r_Datetime_types.html#r_Datetime_types-timestamp

> If you insert a date into a TIMESTAMP column, or a date with a partial time stamp value, the value is implicitly converted into a full time stamp value with default values (00) for missing hours, minutes, and seconds. **Time zone values in input strings are ignored.**
>
> By default, TIMESTAMP values are Coordinated Universal Time (UTC) in both user tables and Amazon Redshift system tables. 

On Nov 3, 2019 Daylight Saving Time Ended at 2:00 AM, when the offset changed from `-0700` to `-0800`.

That means you will have rows in this table where the times area all appearing between `'2019-11-03T01:00:00'` and `'2019-11-03T02:00:00'` where the downloads were, in fact, one hour apart. You won't be able to discern which rows are from the downloads occurring while daylight savings was in effect and when it wasn't.

I don't see this condition handled elsewhere, so I can't approve this yet.

Reply:
Basically, I think you need a new DDL for `asset_downloads_derived` and to not cast as a `TIMESTAMP` here (keep it as a `TIMESTAMPTZ`).